### PR TITLE
Add CREATE Privilege

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

**Review by martint — APPROVED** [review]

**Comment by kokosing** [comment]
What is the semantic of `CREATE`? Can I grant `CREATE` to table actually or is it only for `SCHEMA`?

**Comment by yikf** [comment]
> What is the semantic of `CREATE`? Can I grant `CREATE` to table actually or is it only for `SCHEMA`?

Maybe you have a privilege to  create the specified table under a database?

**Comment by dain** [comment]
> What is the semantic of `CREATE`? Can I grant `CREATE` to table actually or is it only for `SCHEMA`?

There are no explicit semantic to CREATE just like every other privilege in Trino.  This is simply passed through to the backend which handles the grant or revoke.  We expect this will be used for permission systems like PostgreSQL use for controlling who can create tables and schemas, but the backend could use it for anything they want.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HivePrivilegeInfo.java [hunk 1]
```diff
@@ -84,6 +84,9 @@ public HivePrincipal getGrantee()
     public static HivePrivilege toHivePrivilege(Privilege privilege)
     {
         switch (privilege) {
+            case CREATE:
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HivePrivilegeInfo.java:87]: The break makes this a bit hard to follow, since this case is actually redundant. Maybe put it at the end
```java
case CREATE: // not supported by Hive
default:
    throw new IllegalArgumentException("Unexpected privilege: " + privilege);

**Review by electrum — APPROVED** [review]

**Comment by kokosing** [comment]
Can you please add tests for this to io.trino.execution.TestGrantOnSchema and io.trino.execution.TestGrantOnTable?

**Comment by dain** [comment]
> According to https://trino.io/docs/current/sql/grant.html
> 
> > Grant on a table grants the specified privilege on all current and future columns of the table.
> 
> I am don't fully understand how CREATE fits here.

The docs imply a specific implementation and that is simply not how Trino works.  We just hand the grant and revoke to a backend implementation.  That said, I think the docs are fine.  People don't read as closely as you are thinking they well.  If it turns out to be a problem, we can always adjust

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HivePrivilegeInfo.java (hunk 1 above)
Comment (ksobolew) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/HivePrivilegeInfo.java]: Grouping it with `default`, on the other hand, prevents Error Prone from checking for exhaustiveness of the switch

**Comment by kokosing** [comment]
> The docs imply a specific implementation and that is simply not how Trino works.

I am not sure that is the way to work. Trino is providing SQL layer on top of different remote data sources. Even if they are not ANSI SQL nor even SQL compatible we are trying to put a layer via connectors to make the to obey the semantics of SQL. I was under impression that we are doing the same way with access control. Authorization backend might be not RBAC aware at all, but if we are allowing to use SQL to control it, then it should obey the semantics. Otherwise it will be difficult to reason what the expected behavior of GRANT statement would be.

If we are adding `CREATE`, then why are not adding `DROP`. And them what about `ALTER` and others?

Currently in order to run `GRANT CREATE ON TABLE nation ...` it is required that `nation` already should exist. Shouldn't we block that case? It does not make sense.

**Comment by findepi** [comment]
> I was under impression that we are doing the same way with access control.

This is/was certainly the case. I think i remember @electrum asking about grant's *semantics*, which wouldn't make sense in "pass thru whatever" model.

> Currently in order to run `GRANT CREATE ON TABLE nation ...` it is required that `nation` already should exist. Shouldn't we block that case? It does not make sense.

i could imagine one could want to apply it to `CREATE OR REPLACE`, but i'd rather see DROP privilege being tested here.
